### ISSUE TITLE
[Freqtab] Seasonal adjustment for freq table and tab file

### ISSUE THREAD

**Issue by BiancaMorandi** [brightwind-dev/brightwind#334]
This is the same issue raised here #244 but as the method has changed I have opened a new issue 

- should implement similar to MOMM process in https://github.com/brightwind-dev/brightwind/issues/298, i.e. weighting the months by their calendar days. 

Method to implement for deriving weight average freq distribution. Where weight is given by number of days in each calendar month.

- each single month (ie each january) - Derive coverage
- give option to filter out single months with coverage less than a threshold
- raise warning if single month coverage < coverage threshold
- raise error if one or more calendar months are missing -> in this case we can't apply seasonal adjustment
- derive distribution for each calendar month (ie all January)
- total_distribution = apply weight to each calendar month distribution (Jan, Feb, etc) - where weight is number of days of month/total days year - for Feb derive the actual days and derive average number of days (ie 1 Feb 28 days and 1 Feb 29 days - weight by (28+29)/2
- ~~add option to return each single month coverage~~

**Comment by BiancaMorandi** [comment]
Method to implement for deriving weight average freq distribution. Where weight is given by number of days in each calendar month.

- each single month (ie each january) - Derive coverage
- give option to filter out single months with coverage less than a threshold
- raise warning if single month coverage < coverage threshold
- raise error if one or more calendar months are missing -> in this case we can't apply seasonal adjustment
- derive distribution for each calendar month (ie all January)
- total_distribution = apply weight to each calendar month distribution (Jan, Feb, etc) - where weight is number of days of month/total days year - for Feb derive the actual days and derive average number of days (ie 1 Feb 28 days and 1 Feb 29 days - weight by (28+29)/2
- ~~add option to return each single month coverage~~

**Comment by BiancaMorandi** [comment]
@AndyBrightWind can you please review this function and see if results are as you expect?

**Comment by AndyBrightWind** [comment]
> @AndyBrightWind can you please review this function and see if results are as you expect?
> 
> @BiancaMorandi to update Changelog file with summary of changes.

@BiancaMorandi - I've reviewed against other internally used tools and am happy its output is consistent.

**Comment by BiancaMorandi** [comment]
> In parallel with @AndyBrightWind reviewing the actual data that is returned I have some comments.
> 
> @BiancaMorandi
> 
> Is there a reason why the `dist_matrix_by_dir_sector` is not updated for this? Isn't that just the same as `freq_table`?
> 
> Up above in the first post of this issue you write a bulleted description of what the seasonal adjustment does. Why not add a description in the doc string to tell the user what the seasonal adjustment is doing if it was set to True?

@stephenholleran I have updated the docstring to explain what the seasonal adjustment does. I agree that we should allow seasonal adjustment for `dist_matrix_by_dir_sector` but I would prefer to make this update in a separate issue in order to reduce the amount of changes made for this issue.

**Comment by stephenholleran** [comment]
@BiancaMorandi 
I changed the print statement to a warning msg to alert the user that some months have been filtered out however warning messages disappear when run for a second time. I think we need to put the warning message that some months have been removed into the plot itself. Something like the below. We can use the plot title to contain the message. Thoughts?

![image](https://user-images.githubusercontent.com/25622575/189880772-b630704b-bc92-45a1-bdc8-0018dbb13999.png)

**Comment by BiancaMorandi** [comment]
@stephenholleran are you happy to have this title also if the end of the sentence is ''The seasonally adjusted frequency table results might be incorrect.''. The way the warning is implemented allows the end of the sentence to be:

1. 'These months are filtered out for deriving the seasonally adjusted frequency table.'
2. 'The seasonally adjusted frequency table results might be incorrect.'

Point 1 is informative (so ok to have it in a plot that goes to reports), Point 2 is not really what you want to show in a report. Please let me know your thoughts. This change is impacting different functions so I would like to be sure that you are happy with this before making any changes.

**Comment by stephenholleran** [comment]
@BiancaMorandi
Why would 'The seasonally adjusted frequency table results might be incorrect.'? They'd be correct for the months that were included? Showing the months that were not included is all we need to tell the user I thought?

**Comment by BiancaMorandi** [comment]
> @BiancaMorandi Why would 'The seasonally adjusted frequency table results might be incorrect.'? They'd be correct for the months that were included? Showing the months that were not included is all we need to tell the user I thought?

@stephenholleran point 2 in my comment above is raised when coverage_threshold is None or zero (ie user doesn't apply any coverage filter). The goal of this is to warn the user that there are some months where coverage is very low (<80%) but no filter is applied. This specific warning was asked to be implemented during discussion we had 1 year ago. This because results might be incorrect if we use low coverage months (ie if I have ~2 years of data: for 1st January I have only 3 days and for 2nd January I have a full month then using the single 3 days for January will mess up the stats and will give incorrect results). Happy to discuss

**Comment by stephenholleran** [comment]
ok, then lets just expand on that message as it was not clear to me what was going on. This msg is only shown when we think there is an insufficient coverage threshold used i.e. < 0.8.

> 'The seasonally adjusted frequency table results may be incorrect when you use an insufficient data coverage threshold. Some months may have very little data coverage and so may skew the statistics. For the provided dataset the monthly coverage for Jan-2016, May-2016, Nov-2017 is lower than 0.8.'

Thoughts?

**Comment by BiancaMorandi** [comment]
@stephenholleran I have fixed the function to have the text under the figure. See example plots below.

![image](https://user-images.githubusercontent.com/65398058/191325325-42f02772-a9f0-4a4a-ae97-65948fca55d8.png)
![image](https://user-images.githubusercontent.com/65398058/191325373-23bbca9e-f0c3-418b-a02f-8c702ee0a8b8.png)
![image](https://user-images.githubusercontent.com/65398058/191325417-cdf6aa85-cac1-475d-9775-7d8c1c8302d1.png)
![image](https://user-images.githubusercontent.com/65398058/191325461-33050d7c-c547-40d4-bda4-11645c5f3eb4.png)
![image](https://user-images.githubusercontent.com/65398058/191325519-96f79dbe-c904-4d1a-9e4a-fffdf86e2b82.png)
